drivers/gpio/gpio_grgpio2.c

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2021 Cobham Gaisler AB
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Driver for GRLIB GRGPIO revision 2.
+ * - iflag determine pending interrupt.
+ * - interrupt map decides interrupt numer if implemented.
+ * - logic or/and/xor registers used when possible
+ */
+
+#define DT_DRV_COMPAT gaisler_grgpio
+
+#include <drivers/gpio.h>
+#include <spinlock.h>
+#include "gpio_utils.h"
+#include "gpio_grgpio.h"
+
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+LOG_MODULE_REGISTER(gpio_grgpio2);
+
+struct cfg {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	volatile struct grgpio_regs *regs;
+	int interrupt;
+};
+
+struct data {
+	struct gpio_driver_data common;
+	struct k_spinlock lock;
+	sys_slist_t cb;
+	uint32_t imask;
+	uint32_t connected;
+	int irqgen;
+};
+
+static void grgpio_isr(const struct device *dev);
+
+static int pin_configure(const struct device *dev,
+				gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct cfg *cfg = dev->config;
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	uint32_t mask = 1 << pin;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if (flags & GPIO_OUTPUT) {
+		k_spinlock_key_t key;
+
+		/*
+		 * Register operations are atomic, but do the sequence under
+		 * lock so it serializes.
+		 */
+		key = k_spin_lock(&data->lock);
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			regs->output_or = mask;
+		} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+			regs->output_and = ~mask;
+		}
+		regs->dir_or = mask;
+		k_spin_unlock(&data->lock, key);
+	} else {
+		regs->dir_and = ~mask;
+	}
+
+	return 0;
+}
+
+static int port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct cfg *cfg = dev->config;
+	*value = cfg->regs->data;
+	return 0;
```

Comment:
If you happen to update this PR for some other reason you could add one empty line between variable definition block and the function body (even if the function is small).

Reply:
Done